arangod/Aql/ExecutionBlockImpl.h

```diff
@@ -297,24 +297,54 @@ class ExecutionBlockImpl final : public ExecutionBlock {
   auto countShadowRowProduced(AqlCallStack& stack, size_t depth) -> void;
 
  private:
+  /**
+   * @brief The PrefetchTask is used to asynchronously prefetch the next batch
+   * from upstream. Each block holds only a single instance (if any), so each
+   * block can have max one pending async prefetch task. This instance is
+   * created on demand when the first async request is spawned, later tasks can
+   * reuse that instance.
+   * The async task is queued on the global scheduler so it can be picked up by
+   * some worker thread. However, sometimes the original thread might that
+   * created the task might be faster, in which case we don't want to wait
+   * until a worker has picked up the task. Instead, any thread that wants to
+   * process the task has to _claim_ it. This is managed via the task's `state`.
+   * 
+   * Before the task is queued on the scheduler, `state` is set to `Pending`.
+   * When a thread wants to process the task, it must call `tryClaim` which
+   * sets `state` to `InProgress` iff it is still pending. If `tryClaim`
+   * returns true, the calling thread is now the owner. If a worker thread
+   * successfully claimed the task, it executes it, stores the result, sets
+   * `state` to `Finished` and signals the `bell`. If it fails to claim the
+   * task, it can immediately drop the task.
+   * If the original thread successfully claimed the task, it can continue as
+   * usual (i.e., as if we have never spawned the task in the first place). If
+   * it fails to claim the task, it has to wait until the worker sets `state`
+   * to `Finished`, then fetches the result and sets `state` to `Consumed`.
+   * This is necessary to avoid waiting for a task that has already been
+   * consumed.
+   */
   struct PrefetchTask {
     enum class State {
       Pending,
       InProgress,
       Finished,
       Consumed
     };
-    std::atomic<State> state{State::Pending};
-
-    std::mutex lock;
-    std::condition_variable bell;
-    std::optional<std::tuple<ExecutionState, SkipResult, typename Fetcher::DataRange>> result;
+    using Result = std::tuple<ExecutionState, SkipResult, typename Fetcher::DataRange>;
     
     bool isConsumed() const noexcept;
     bool tryClaim() noexcept;
     void waitFor() noexcept;
+    void reset() noexcept;
+    Result stealResult() noexcept;
     
     void execute(ExecutionBlockImpl& block, AqlCallStack& stack);
+    
+   private:
+    std::atomic<State> _state{State::Pending};
+    std::mutex _lock;
+    std::condition_variable _bell;
+    std::optional<Result> _result;
```

Comment:
The typedef `Result` here can be a little bit misleading, thinking of our `arangodb::Result` type.